biogtr/models/model_utils.py

```diff
@@ -0,0 +1,67 @@
+"""Module containing model helper functions."""
+from typing import Dict, List, Tuple
+import torch
+
+
+def get_boxes_times(instances: List[Dict]) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Extracts the bounding boxes and frame indices from the input list of instances.
+
+    Args:
+        instances (List[Dict]): List of instance dictionaries
+
+    Returns:
+        Tuple[torch.Tensor, torch.Tensor]: A tuple of two tensors containing the
+                                            bounding boxes and corresponding frame
+                                            indices, respectively.
+    """
+    boxes, times = [], []
+    _, h, w = instances[0]["img_shape"].flatten()
+
+    for fidx, instance in enumerate(instances):
+        bbox = instance["bboxes"]
+        bbox[:, [0, 2]] /= w
+        bbox[:, [1, 3]] /= h
+
+        boxes.append(bbox)
+        times.append(torch.full((bbox.shape[0],), fidx))
+
+    boxes = torch.cat(boxes, dim=0)  # N x 4
+    times = torch.cat(times, dim=0).to(boxes.device)  # N
+    return boxes, times
+
+
+def softmax_asso(asso_output: list[torch.Tensor]) -> list[torch.Tensor]:
+    """Applies the softmax activation function on asso_output.
+
+    Args:
+        asso_output: Same structure as before. It's a list of tensors.
+        An example is shown  below. The shape is modified.
+
+    Returns:
+        asso_output: Exactly the same as before but with the softmax applied.
+        # ------------------------ An example of asso_output ----------------------- #
+        N_i: number of detected instances in i-th frame of window.
+        N_t: number of instances in current/query frame (rightmost frame of the window).
+        T: length of window.
+        asso_output is of shape: (T, N_t, N_i).
+    """
+    # N_i: number of detected instances in i-th frame of window.
+    # N_t: number of instances in current frame (rightmost frame of the window).
+    # T: length of window.
+
+    # asso_output: (T, N_t, N_i)
+
+    asso_active = []
+    for asso in asso_output:
+        # asso: (N_t, N_i)
+
+        # I'm guessing what they are doing here is giving the model a chance to pick "uncertain".
```

Comment:
can remove this comment. this function is just to get probs from logits during inference.